### ISSUE TITLE
File lock issue on Cray system? (fcntl.lockf : Bad file descriptor)

### ISSUE THREAD

**Issue by pramodk** [LLNL/spack#1904]
Trying spack first time on Cray system and see following:

```bash
$ git clone https://github.com/LLNL/spack.git
$ ./spack/bin/spack arch
cray-cnl-sahdybridge
```

But for find command:
```
$ ./spack/bin/spack find
Traceback (most recent call last):
  File "./spack/bin/spack", line 192, in <module>
    main()
  File "./spack/bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/users/kumbhar/spack/lib/spack/spack/cmd/find.py", line 148, in find
    specs = set(spack.installed_db.query(**q_args))
  File "/users/kumbhar/spack/lib/spack/spack/database.py", line 630, in query
    with self.read_transaction():
  File "/users/kumbhar/spack/lib/spack/llnl/util/lock.py", line 211, in __enter__
    self._as = self._acquire_fn()
        self._file_path = file_path
  File "/users/kumbhar/spack/lib/spack/spack/database.py", line 447, in _read
    self.reindex(spack.install_layout)
  File "/users/kumbhar/spack/lib/spack/spack/database.py", line 351, in reindex
    with transaction:
  File "/users/kumbhar/spack/lib/spack/llnl/util/lock.py", line 210, in __enter__
    if self._enter() and self._acquire_fn:
  File "/users/kumbhar/spack/lib/spack/llnl/util/lock.py", line 241, in _enter
    return self._lock.acquire_write(self._timeout)
  File "/users/kumbhar/spack/lib/spack/llnl/util/lock.py", line 134, in acquire_write
    self._lock(fcntl.LOCK_EX, timeout)   # can raise LockError.
  File "/users/kumbhar/spack/lib/spack/llnl/util/lock.py", line 76, in _lock
    fcntl.lockf(self._fd, op | fcntl.LOCK_NB)
IOError: [Errno 9] Bad file descriptor
```

I see this issue on TITAN and Daint Cray systems.
Any suggestion?

**Comment by tgamblin** [comment]
Fixed by #1906

**Comment by pramodk** [comment]
I see similar issue on OSX in below scenario:
* Clone new repo 
* No packages are installed 
* Try to find packages as
```bash
$ git clone https://github.com/llnl/spack.git spack-llnl
Cloning into 'spack-llnl'...
remote: Counting objects: 44080, done.
................

 $ cd spack-llnl/

$ ./bin/spack find
Traceback (most recent call last):
  File "./bin/spack", line 192, in <module>
    main()
  File "./bin/spack", line 169, in main
    return_val = command(parser, args)
  File "/Users/kumbhar//spack-llnl/lib/spack/spack/cmd/find.py", line 148, in find
    specs = set(spack.installed_db.query(**q_args))
  File "/Users/kumbhar//spack-llnl/lib/spack/spack/database.py", line 630, in query
    with self.read_transaction():
  File "/Users/kumbhar//spack-llnl/lib/spack/llnl/util/lock.py", line 219, in __enter__
    self._as = self._acquire_fn()
  File "/Users/kumbhar//spack-llnl/lib/spack/spack/database.py", line 447, in _read
    self.reindex(spack.install_layout)
  File "/Users/kumbhar//spack-llnl/lib/spack/spack/database.py", line 351, in reindex
    with transaction:
  File "/Users/kumbhar//spack-llnl/lib/spack/llnl/util/lock.py", line 218, in __enter__
    if self._enter() and self._acquire_fn:
  File "/Users/kumbhar//spack-llnl/lib/spack/llnl/util/lock.py", line 249, in _enter
    return self._lock.acquire_write(self._timeout)
  File "/Users/kumbhar//spack-llnl/lib/spack/llnl/util/lock.py", line 142, in acquire_write
    self._lock(fcntl.LOCK_EX, timeout)   # can raise LockError.
  File "/Users/kumbhar//spack-llnl/lib/spack/llnl/util/lock.py", line 84, in _lock
    fcntl.lockf(self._fd, op | fcntl.LOCK_NB)
IOError: [Errno 9] Bad file descriptor
```
* Once you install first package, this doesn't appear.
* Not on linux/other platforms
